File: examples/python/failure/record.py
"""Capture a failure; exception messages are not automatically recorded."""

from pathlib import Path

import refract

Path(".examples").mkdir(exist_ok=True)

try:
    with refract.run("failed-lookup", path=".examples/failure.rfr", fail_open=False):
        refract.event(
            type="tool.call",
            name="Lookup customer",
            input={"api_key": "demo-secret"},
            output={"found": False},
        )
        raise LookupError("No matching customer")
except LookupError:
    print("Wrote .examples/failure.rfr with failed status and a redacted input")

File: examples/typescript/concurrent/record.mjs
import { mkdir } from "node:fs/promises";
await mkdir(".examples", { recursive: true });
import { refract } from "../../../packages/typescript/dist/index.js";
await Promise.all(
  ["alpha", "beta"].map((name) =>
    refract.run(
      name,
      async () => {
        await Promise.resolve();
        refract.event({
          type: "tool.call",
          name: "lookup",
          input: { api_key: "demo-secret" },
          output: { team: name },
        });
      },
      { path: `.examples/${name}.rfr` },
    ),
  ),
);
